File: minmax/guess_the_word.rb
#    This is Master's API interface.
#    You should not implement it, or speculate about its implementation
#
# class Master
# =begin
#     :type word: String
#     :rtype: Integer
# =end
#     def guess(word)
#         ...
#     end
# end
#

# Approach 1: Reduce wordlist to words that match with first guess
# @param {String[]} wordlist
# @param {Master} master
# @return {Void}
def find_secret_word(wordlist, master)
    n = 0
    while n < 6
        guess = wordlist.sample
        n = master.guess(guess)
        wordlist = wordlist.select { |w| match(w, guess) == n }
    end
end

def match(w1, w2)
    w1.chars.zip(w2.chars).map { |c1, c2| c1 == c2 ? 1 : 0 }.reduce(0, :+)
end

# Approach 2: MinMax
# @param {String[]} wordlist
# @param {Master} master
# @return {Void}
def find_secret_word(wordlist, master)
    n = 0
    while n < 6
        counts = Hash.new { |h, k| h[k] = 0 }
        wordlist.permutation(2).map { |w1, w2| w1 if match(w1, w2) == 0 }.each { |w| counts[w] += 1 }
        guess = wordlist.min_by { |w| counts[w] }
        n = master.guess(guess)
        wordlist = wordlist.select { |w| match(w, guess) == n }
    end
end


def match(w1, w2)
    w1.chars.zip(w2.chars).map { |c1, c2| c1 == c2 ? 1 : 0 }.reduce(0, :+)
end



# Approach 1: Reduce wordlist by eliminating using match size
# Steps:
# 1. Loop till match size == 6
# 2. Take a random choice from list and capture match size
# 3. Now only select words from wordlist that have similar
#    match size as they are candidates for secret.
# Time: O(n)
# Space: O(1)


# Approach 2: MinMax
# Steps:
# 1. Choose a word that has closest distance to the maximum number of words in wordlist
# 2. Use that as the first guess and get the distance from secret.
# 3. We get to eliminate all the words that don't have that distance,
#    because they can't be candidates as they don't have
#    the same distance from the first guess word as secret.
# 4. We continue the loop till we get a correct guess,
#    eliminating words and choosing the closest word,
# Time: O(n^2)
# Space: O(n^2)

File: minmax/guess_the_word.py
from typing import List

class Master:
    def guess(self, word: str) -> int:
        pass

# Approach 2: heuristic 1
import collections
import itertools
class Solution:
    def findSecretWord(self, wordlist: List[str], master: 'Master'):
        n = 0
        while n < 6:
            count = collections.Counter(w1 for w1, w2 in itertools.permutations(wordlist, 2) if self.match(w1, w2) == 0)
            guess = min(wordlist, key=lambda w: count[w])
            self.print_counter(count)
            n = master.guess(guess)
            wordlist = [w for w in wordlist if self.match(w, guess) == n]

    def print_counter(self, count):
        for key, value in count.items():
            print(key, value)

    def match(self, word1, word2):
        count = 0
        for c1, c2 in zip(word1, word2):
            if c1 == c2: count += 1
        return count

# Approach 2: heuristic 2

class Solution:
    def findSecretWord(self, wordlist: List[str], master: 'Master'):
        while wordlist:
            s = self.most_overlap_word(wordlist)     # guess the word that overlaps with most others
            n = master.guess(s)
            if n == 6: return
            # wordlist = [w for w in wordlist if self.pair_matches(s, w) == n]   # filter words with same matches
            wordlist = list(filter(lambda w: self.pair_matches(s, w) == n, wordlist))

    def pair_matches(self, a, b):                   # count the number of matching characters
        return sum(c1 == c2 for c1, c2 in zip(a, b))

    def most_overlap_word(self, candidates):
        counts = [collections.defaultdict(int) for _ in range(6)]

        for word in candidates:
            for i, c in enumerate(word):
                counts[i][c] += 1

        return max(candidates, key=lambda w: sum(counts[i][c] for i, c in enumerate(w)))


# 843. Guess the Word
# https://leetcode.com/problems/guess-the-word/

# Approach 1: Simple Counting and Guessing


# Approach 2: MinMax with heuristic
# heuristic 1: Least match
# heuristic 2: most overlap

# Steps:
# 1. Choose a word that has most overlap [highest matches with hig]
# 2. Use that as the first guess and get the distance from secret.
# 3. We get to eliminate all the words that don't have that distance,
#    because they can't be candidates as they don't have
#    the same distance from the first guess word as secret.
# 4. We continue the loop till we get a correct guess,
#    filtering words that are the at the distance of the bad guesss
#    and choosing the next most overlap word

# Time: O(n)
# Space: O(1), re-use wordList
